qa/rolling-upgrade/src/test/java/org/opensearch/upgrades/IndexingIT.java

```diff
@@ -88,52 +83,46 @@ private void printClusterNodes() throws IOException, ParseException, URISyntaxEx
     }
 
     // Verifies that for each shard copy holds same document count across all containing nodes.
-    private void waitForSearchableDocs(String index, int shardCount) throws Exception {
-        Map<Integer,String> primaryShardToNodeIDMap = new HashMap<>();
-        Map<Integer,String> replicaShardToNodeIDMap = new HashMap<>();
+    private void waitForSearchableDocs(String index, int shardCount, int replicaCount) throws Exception {
+        assertTrue(shardCount > 0);
+        assertTrue(replicaCount > 0);
         waitForClusterHealthWithNoShardMigration(index, "green");
         logger.info("--> _cat/shards before search \n{}", EntityUtils.toString(client().performRequest(new Request("GET", "/_cat/shards?v")).getEntity()));
 
+        // Verify segment replication stats
         verifySegmentStats(index);
-        Request request = new Request("GET", index + "/_stats");
-        request.addParameter("level", "shards");
-        Response response = client().performRequest(request);
-        for (int shardNumber = 0; shardNumber < shardCount; shardNumber++) {
-            List<Object> shardStats = ObjectPath.createFromResponse(response).evaluate("indices." + index + ".shards." + shardNumber);
-            for (Object shard : shardStats) {
-                final String nodeId = ObjectPath.evaluate(shard, "routing.node");
-                final Boolean primary = ObjectPath.evaluate(shard, "routing.primary");
-                if (primary) {
-                    primaryShardToNodeIDMap.putIfAbsent(shardNumber, nodeId);
-                } else {
-                    replicaShardToNodeIDMap.putIfAbsent(shardNumber, nodeId);
+
+        // Verify segment store
+        assertBusy(() -> {
+            Request segrepStatsRequest = new Request("GET", "/_cat/segments/" + index + "?s=shard,segment,primaryOrReplica");
+            segrepStatsRequest.addParameter("h", "index,shard,primaryOrReplica,segment,docs.count");
+            Response segrepStatsResponse = client().performRequest(segrepStatsRequest);
+            logger.info("--> _cat/segments response\n {}", EntityUtils.toString(segrepStatsResponse.getEntity()));
+            List<String> responseList = Streams.readAllLines(segrepStatsResponse.getEntity().getContent());
+            for (int segmentsIndex=0; segmentsIndex < responseList.size();) {
+                String[] primaryRow = responseList.get(segmentsIndex++).split(" +");
```

Comment:
nit: Can we use the `ObjectPath.createFromResponse` pattern to make it more readable? Comments here help, but I think it would be easier to get context with path based access.

Reply:
Yes, `ObjectPath.createFromResponse` can be used. Though I prefer existing format for below reasons:
1. I need to process entire output vs specific fields inside nested object. 
2. I am logging the response in default format (which is compact and widely used), thus keeping logic to parse same response will be helpful during debugging. 

For comparison and completeness, below is the sample response (truncated for brevity). 
Default
```
 1>  index                            shard prirep ip        segment generation docs.count docs.deleted  size size.memory committed searchable version compound
  1> auto_id_and_op_type_create_index 0     p      127.0.0.1 _0               0          1            0 3.5kb           0 false     true       9.6.0   true
  1> index_with_replicas              0     p      127.0.0.1 _0               0          5            0 4.2kb           0 true      true       9.6.0   true
  1> test_index                       0     p      127.0.0.1 _0               0          5            0 4.2kb           0 false     true       9.6.0   true
  1> test-index-segrep                0     p      127.0.0.1 _0               0          2            0 3.9kb           0 false     true       9.6.0   true
```

JSON
```
  1> [2023-06-28T00:43:48,596][INFO ][o.o.u.IndexingIT         ] [testIndexingWithSegRep] --> _cat/segments response
  1>  [
  1>   {
  1>     "index" : "auto_id_and_op_type_create_index",
  1>     "shard" : "0",
  1>     "prirep" : "p",
  1>     "ip" : "127.0.0.1",
  1>     "segment" : "_0",
  1>     "generation" : "0",
  1>     "docs.count" : "1",
  1>     "docs.deleted" : "0",
  1>     "size" : "3.5kb",
  1>     "size.memory" : "0",
  1>     "committed" : "false",
  1>     "searchable" : "true",
  1>     "version" : "9.6.0",
  1>     "compound" : "true"
  1>   },
  1>   {
  1>     "index" : "index_with_replicas",
  1>     "shard" : "0",
  1>     "prirep" : "p",
  1>     "ip" : "127.0.0.1",
  1>     "segment" : "_0",
  1>     "generation" : "0",
  1>     "docs.count" : "5",
  1>     "docs.deleted" : "0",
  1>     "size" : "4.2kb",
  1>     "size.memory" : "0",
  1>     "committed" : "true",
  1>     "searchable" : "true",
  1>     "version" : "9.6.0",
  1>     "compound" : "true"
  1>   },
```